src/mongocxx/lib/mongocxx/v1/pipeline.cpp

```diff
@@ -12,4 +12,292 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#include <mongocxx/v1/pipeline.hpp>
+#include <mongocxx/v1/pipeline.hh>
+
+//
+
+#include <bsoncxx/v1/array/view.hpp>
+#include <bsoncxx/v1/document/view.hpp>
+#include <bsoncxx/v1/stdx/string_view.hpp>
+#include <bsoncxx/v1/types/view.hpp>
+
+#include <array>
+#include <cinttypes>
+#include <cstdint>
+#include <cstdio>
+#include <string>
+
+#include <bsoncxx/private/bson.hh>
+
+#include <mongocxx/private/scoped_bson.hh>
+#include <mongocxx/private/utility.hh>
+
+namespace mongocxx {
+namespace v1 {
+
+class pipeline::impl {
+   private:
+    static_assert(INT32_MAX == std::int32_t{2147483647}, "");
+    std::array<char, sizeof("2147483647")> _idx = {}; // Access via `this->idx()`.
+
+   public:
+    scoped_bson _doc;
+    std::int32_t _count = 0;
+
+    char const* idx() {
+        (void)std::snprintf(_idx.data(), _idx.size(), "%" PRId32, _count);
+        return _idx.data();
+    }
+
+    void append(bsoncxx::v1::document::view doc) {
+        _doc += scoped_bson{BCON_NEW(this->idx(), BCON_DOCUMENT(scoped_bson_view{doc}.bson()))};
+        ++_count;
```

Comment:
Aside: I expect the unchecked `++_count` is OK. If a caller were to append INT32_MAX stages, `_doc` would sooner overflow an result in an exception when appending to `_doc`.